lorawan-device/src/region/dynamic_channel_plans/mod.rs

```diff
@@ -144,40 +144,63 @@ impl<R: DynamicChannelRegion> RegionHandler for DynamicChannelPlan<R> {
         }
     }
 
-    fn handle_link_adr_channel_mask(
-        &mut self,
-        channel_mask_control: u8,
-        channel_mask: ChannelMask<2>,
-    ) {
-        match channel_mask_control {
+    fn channel_mask_get(&self) -> ChannelMask<9> {
+        self.channel_mask.clone()
+    }
+
+    fn channel_mask_set(&mut self, channel_mask: ChannelMask<9>) {
+        self.channel_mask = channel_mask;
+    }
+
+    fn channel_mask_update(
+        &self,
+        channel_mask: &mut ChannelMask<9>,
+        ch_mask_ctl: u8,
+        ch_mask: ChannelMask<2>,
+    ) -> Option<()> {
+        match ch_mask_ctl {
             0..=4 => {
-                let base_index = channel_mask_control as usize * 2;
-                self.channel_mask.set_bank(base_index, channel_mask.get_index(0));
-                self.channel_mask.set_bank(base_index + 1, channel_mask.get_index(1));
+                let base_index = ch_mask_ctl as usize * 2;
+                channel_mask.set_bank(base_index, ch_mask.get_index(0));
+                channel_mask.set_bank(base_index + 1, ch_mask.get_index(1));
             }
             5 => {
-                let channel_mask: u16 =
-                    channel_mask.get_index(0) as u16 | ((channel_mask.get_index(1) as u16) << 8);
-                self.channel_mask.set_bank(0, ((channel_mask & 0b1) * 0xFF) as u8);
-                self.channel_mask.set_bank(1, ((channel_mask & 0b10) * 0xFF) as u8);
-                self.channel_mask.set_bank(2, ((channel_mask & 0b100) * 0xFF) as u8);
-                self.channel_mask.set_bank(3, ((channel_mask & 0b1000) * 0xFF) as u8);
-                self.channel_mask.set_bank(4, ((channel_mask & 0b10000) * 0xFF) as u8);
-                self.channel_mask.set_bank(5, ((channel_mask & 0b100000) * 0xFF) as u8);
-                self.channel_mask.set_bank(6, ((channel_mask & 0b1000000) * 0xFF) as u8);
-                self.channel_mask.set_bank(7, ((channel_mask & 0b10000000) * 0xFF) as u8);
-                self.channel_mask.set_bank(8, ((channel_mask & 0b100000000) * 0xFF) as u8);
+                let ch_mask: u16 =
+                    ch_mask.get_index(0) as u16 | ((ch_mask.get_index(1) as u16) << 8);
+                channel_mask.set_bank(0, ((ch_mask & 0b1) * 0xFF) as u8);
+                channel_mask.set_bank(1, ((ch_mask & 0b10) * 0xFF) as u8);
+                channel_mask.set_bank(2, ((ch_mask & 0b100) * 0xFF) as u8);
+                channel_mask.set_bank(3, ((ch_mask & 0b1000) * 0xFF) as u8);
+                channel_mask.set_bank(4, ((ch_mask & 0b10000) * 0xFF) as u8);
+                channel_mask.set_bank(5, ((ch_mask & 0b100000) * 0xFF) as u8);
+                channel_mask.set_bank(6, ((ch_mask & 0b1000000) * 0xFF) as u8);
+                channel_mask.set_bank(7, ((ch_mask & 0b10000000) * 0xFF) as u8);
+                channel_mask.set_bank(8, ((ch_mask & 0b100000000) * 0xFF) as u8);
```

Comment:
I know you didn't write it originally and you only modified it, but perhaps:
```suggestion
                for bank in 0..9 {
                    let bit_set = (ch_mask >> bank) & 1;
                    channel_mask.set_bank(bank, (bit_set * 0xFF) as u8);
                }
```

Reply:
Will be done as future improvement as there's also some controversy in the docs regarding `ch_mask_ctl == 5`  with follow up errata.